gix-blame/src/lib.rs

```diff
@@ -2,9 +2,960 @@
+        } else {
+            acc.push(entry);
+
+            acc
+        }
+    })
+}
+
+// TODO: do not instantiate anything, get everything passed as argument.
+pub fn blame_file<E>(
+    odb: impl gix_object::Find + gix_object::FindHeader,
+    traverse: impl IntoIterator<Item = Result<gix_traverse::commit::Info, E>>,
+    resource_cache: &mut gix_diff::blob::Platform,
+    suspect: ObjectId,
+    worktree_path: PathBuf,
+    file_path: &BStr,
+) -> Result<Vec<BlameEntry>, E> {
+    // TODO
+    // At a high level, what we want to do is the following:
+    //
+    // - get the commit that belongs to a commit id
+    // - walk through parents
+    //   - for each parent, do a diff and mark lines that don’t have a suspect (this is the term
+    //     used in `libgit2`) yet, but that have been changed in this commit
+    //
+    // The algorithm in `libgit2` works by going through parents and keeping a linked list of blame
+    // suspects. It can be visualized as follows:
+    //
+    // <---------------------------------------->
+    // <---------------><----------------------->
+    // <---><----------><----------------------->
+    // <---><----------><-------><-----><------->
+    // <---><---><-----><-------><-----><------->
+    // <---><---><-----><-------><-----><-><-><->
+
+    // Needed for `to_str`.
+    use gix_object::bstr::ByteSlice;
+
+    let absolute_path = worktree_path.join(file_path.to_str().unwrap());
+
+    // TODO Verify that `imara-diff` tokenizes lines the same way `lines` does.
+    let number_of_lines = std::fs::read_to_string(absolute_path).unwrap().lines().count();
+
+    let mut hunks_to_blame: Vec<UnblamedHunk> = vec![UnblamedHunk::new(
+        0..number_of_lines.try_into().unwrap(),
+        suspect,
+        Offset::Added(0),
+    )];
+    let mut out: Vec<BlameEntry> = vec![];
+
+    'outer: for item in traverse {
+        let item = item?;
+        let suspect = item.id;
+
+        let parent_ids = item.parent_ids;
+        if parent_ids.is_empty() {
+            // I’m not entirely sure if this is correct yet. `suspect`, at this point, is the `id` of
+            // the last `item` that was yielded by `traverse`, so it makes sense to assign the
+            // remaining lines to it, even though we don’t explicitly check whether that is true
+            // here. We could perhaps use `needed_to_obtain` to compare `suspect` against an empty
+            // tree to validate this assumption.
+            out.extend(
+                hunks_to_blame
+                    .iter()
+                    .map(|hunk| BlameEntry::from_unblamed_hunk(hunk, suspect)),
+            );
+
+            hunks_to_blame = vec![];
+
+            break;
+        }
+
+        let mut buffer = Vec::new();
+        let commit_id = odb.find_commit(&suspect, &mut buffer).unwrap().tree();
+        let tree = odb.find_tree(&commit_id, &mut buffer).unwrap();
+
+        let Some(entry) = tree.bisect_entry(file_path, false) else {
```

Comment:
@Byron Given [your comment][comment-lookup-entry], it seems like `bisect_entry` might be the wrong choice here as well. `gix_object` doesn’t seem to provide a method comparable to `lookup_entry`, though. Am I missing something?

[comment-lookup-entry]: https://github.com/GitoxideLabs/gitoxide/pull/1643#issuecomment-2439386952

Reply:
Right, the lower-level cousin doesn't support nested lookup, but there is no reason it shouldn't be able to do that. 
It's probably possible to port `lookup_path()` from `gix` down to `gix-object`, or `gix-blame` implements its own version of it.

Reply:
That's great to hear, thanks a lot!

Let's have it in a separate PR so I can merge it much more quickly.

Reply:
New PR: #1686.